orm/model_bucket.go

```diff
@@ -68,16 +68,47 @@ type ModelBucket interface {
 // NewModelBucket returns a ModelBucket instance. This implementation relies on
 // a bucket instance. Final implementation should operate directly on the
 // KVStore instead.
-func NewModelBucket(b Bucket) ModelBucket {
-	return &modelBucket{
+func NewModelBucket(name string, m Model, opts ...ModelBucketOption) ModelBucket {
+	b := NewBucket(name, NewSimpleObj(nil, m))
+
+	tp := reflect.TypeOf(m)
+	if tp.Kind() == reflect.Ptr {
+		tp = tp.Elem()
+	}
+
+	mb := &modelBucket{
 		b:     b,
 		idSeq: b.Sequence("id"),
+		model: tp,
+	}
+	for _, fn := range opts {
+		fn(mb)
+	}
+	return mb
+}
+
+// ModelBucketOption is implemented by any function that can configure
+// ModelBucket during creation.
+type ModelBucketOption func(mb *modelBucket)
+
+// WithIndex configures the bucket to build an index with given name. All
+// entities stored in the bucket are indexed using value returned by the
+// indexer function. If an index is unique, there can be only one entity
+// referenced per index value.
+func WithIndex(name string, indexer Indexer, unique bool) ModelBucketOption {
```

Comment:
Okay, as I understand it, the following are the same:

`b := orm.NewModelBucket("foo", &Model{}, WithIndex("count", countFunc, false))`

```golang
b := orm.NewModelBucket("foo", &Model{})
WithIndex("count", countFunc, false)(b)
```